core/container/array/array.odin

```diff
@@ -0,0 +1,189 @@
+/*
+This is an array that uses a fixed array as backing. The capacity of the array
```

Comment:
I'd move the documentation of what it does over to the type. Also I don't like the wording usage of "this is X", because it may risk being unspecific.

Though, saying "An Array is an array that does [...]" also sounds strange, because it conflates with the builtin array type. I'm afraid that that is a seed for potential communication issues. Which is maybe why I would advocate for a different name, other than an Array.

I agree that the previous name, `Small_Array` was weird. There's nothing in it to make it small, or at least smaller than a normal array. I also don't like the options you listed as alternatives in your PR description.

I have a few more suggestions, `Growing_Array`, or maybe something that suggests a list, like `Fixed_List`. Right now my suggestions aren't good, I'd have to look more into existing naming schemes. Also to note, the way this container functions is almost exactly like a dynamic array, equipped with `nil` allocator. Maybe something to think about

Reply:
I welcome other name suggestions.

The weird problem with naming this thing is that I always feel like the name either has possible communication issues or that I disregard looking into the package because the name sounds weird.

I think at least the simple name `Array` makes me more likely to open and read the docs instead of disregarding it like I did when it had a strange name. And if I open it and read the docs and see that it does not do what I want from an array, then I can look elsewhere.

Reply:
Btw, regarding this:
> Also to note, the way this container functions is almost exactly like a dynamic array, equipped with nil allocator. Maybe something to think about
I think the Small_Array style of array is popular with the never-do-dynamic-allocations-crowd. I guess one could maybe setup the dynamic array to point to fixed memory though.

Reply:
> I welcome other name suggestions.

`Inline_Array`, in that the data is inline in the struct, not a pointer to the data.

Reply:
I like `Inline_Array`.

Reply:
> I welcome other name suggestions.

I don't like calling it just `Array`; that's a significant name overlap with the concept of arrays. My first thought went to `Limited_Array`, but all arrays are technically limited if they're not dynamic. `Stack_Array` (stack as in heap/stack, not the datatype) has the potential to be confused with the idea of stack datatypes (and the struct could be allocated, too), but it _does_ line up well with some of the API (`push_back`, `pop_front`).

`Inline_Array` might be the better option.

Reply:
The convention I'm used to for this type of array, one that grows within its memory budget, is `Fixed_Array`. I do like `Inline_Array` though as that makes it easy to intuit where the storage for the array is.

Reply:
`Inline_Array` sounds good to me. I set this PR as draft and will change to that name when I get some time.

Reply:
@flysand7 I kept the documentation at the top of the file. It feels more like an overview of both the type but also what the procedures in there do. However, I'm not sure if there is some official policy here, and I can change it if there is. Personally I really like overview comments at the top of files.

Reply:
Since its in its own separate package I guess either way would work fine. But generally speaking, I put documentation into the place where it applies the most. If it describes the functionality of the package as a whole, like why this package is here and what kinds of concepts we'll be using to document this packages' functions, it belongs to the package documentation. Otherwise it belongs to the type.

For example, with `core:mem` docs I put the general concepts like pointer alignment, memory allocation, ZII into the package docs, because I will use these concepts and I felt it would be needed to be described somewhere that's not "under every single function in the package". I've also had trouble deciding whether the `Arena` type is the one I'm gonna put the arena allocators' documentation to, or is it going to be the `arena_allocator()` function. At the end I decided on the latter.

But that was a package with more than one thing in it, so, I don't know. It would be fine either way, but we should decide on our measure of consistency for this, because we can't have packages spread out in how they document themselves while implementing only one concept.

Reply:
That make sense. I've written big systems for game engines in the past, in those cases I really like a big overview comment at the top of the most central file. It's often the only documentation I need to read and it's the first thing I see.